streaming/text/convert/pile.py

```diff
@@ -0,0 +1,150 @@
+        in_files (List[str]): List of input files to generate arguments for.
+
+    Returns:
+        Iterator[Tuple[str, str, str, List[str], int]]: Each argument tuple.
+    """
+    for in_file in in_files:
+        assert in_file.startswith(in_root)
+        assert in_file.endswith('.jsonl')
+        out_dir = os.path.join(out_root, in_file[len(in_root):-len('.jsonl')])
+        yield in_file, out_dir, compression, hashes, size_limit
+
+
+def file_to_dir(args: Tuple[str, str, str, List[str], int]) -> Dict[str, int]:
+    """Convert a JSONL input file into a directory of MDS shards.
+
+    This is the unit of work executed by the process pool.
+
+    Args:
+        args (Tuple[str, str, str, List[str], int): All arguments, packed into a tuple because
+            process pools only pass one argument.
+
+    Returns:
+        Dict[str, int]: Count of how many samples belonged to each Pile dataset subset.
+    """
+    in_file, out_dir, compression, hashes, size_limit = args
+
+    columns = {
+        'text': 'str',
+        'pile_set_name': 'str',
+    }
+
+    counts = Counter()
+    with MDSWriter(out_dir, columns, compression, hashes, size_limit) as out:
+        for line in open(in_file):
+            obj = json.loads(line)
+            if sorted(obj.keys()) != ['meta', 'text']:
+                raise ValueError('Invalid sample fields.')
+            text = obj['text']
+            meta = obj['meta']
+            if sorted(meta.keys()) != ['pile_set_name']:
+                raise ValueError('Invalild sample meta fields.')
+            pile_set_name = meta['pile_set_name']
+            sample = {
+                'text': text,
+                'pile_set_name': pile_set_name,
+            }
+            out.write(sample)
+            counts[pile_set_name] += 1
+    return counts
+
+
+def main(args: Namespace) -> None:
+    """Convert the Pile to streaming format.
+
+    Args:
+        args (Namespace): Command-line arguments.
+    """
+    hashes = get_list_arg(args.hashes)
+
+    # Find the original JSONL files to convert.
+    pattern = os.path.join(args.in_root, 'train', '*.jsonl')
+    trains = sorted(glob(pattern))
+    val = os.path.join(args.in_root, 'val.jsonl')
+    test = os.path.join(args.in_root, 'test.jsonl')
+    in_files = trains + [val, test]
+
+    # Get the arguments for each JSONL file conversion.
+    arg_tuples = each_task(args.in_root, args.out_root, args.compression, hashes, args.size_limit,
+                           in_files)
+
+    # Process each JSONL file in parallel into directories of shards.
+    with Pool() as pool:
+        counters = pool.imap(file_to_dir, arg_tuples)
+        for in_file, counts in zip(in_files, counters):
+            obj = {
+                'file': in_file,
+                'counts': counts,
+            }
+            print(json.dumps(obj, sort_keys=True))
+
+
+if __name__ == '__main__':
+    main(parse_args())
```

Comment:
Can you put a global comment on how one should download the raw pile dataset ? and how the directory structure looks like ? Otherwise, it would be confusing for the people on how this script!!

Reply:
Fixed